File: core/src/main/kotlin/org/bitlap/core/sql/PrunePushedFilter.kt
/* Copyright (c) 2022 bitlap.org */
package org.bitlap.core.sql

import java.io.Serializable

/**
 * Prune other filter to push down
 */
typealias PushedFilterFun = (String) -> Boolean

data class PrunePushedFilterExpr(
    val name: String,
    val op: FilterOp,
    val values: List<String>,
    val func: PushedFilterFun,
    val expr: String,
)

open class PrunePushedFilter : Serializable {

    private val conditions = mutableListOf<PrunePushedFilterExpr>()

    fun add(name: String, op: FilterOp, values: List<String>, func: PushedFilterFun, expr: String): PrunePushedFilter {
        return this.also { it.conditions.add(PrunePushedFilterExpr(name, op, values, func, expr)) }
    }

    fun filter(name: String?): PrunePushedFilter {
        val rs = this.conditions.filter { it.name == name }
        return PrunePushedFilter().also {
            it.conditions.addAll(rs)
        }
    }

    fun getNames(): Set<String> = this.conditions.map { it.name }.toSet()
    fun getConditions(): List<PrunePushedFilterExpr> = this.conditions

    override fun toString(): String {
        return this.conditions.map { it.expr }.toString()
    }
}

File: bitlap-core/src/main/scala/org/bitlap/core/sql/PrunePushedFilter.scala
/**
 * Copyright (C) 2023 bitlap.org .
 */
package org.bitlap.core.sql

import java.io.Serializable

import scala.collection.mutable.ListBuffer

/** Prune other filter to push down
 */
type PushedFilterFun = (String) => Boolean

final case class PrunePushedFilterExpr(
  name: String,
  op: FilterOp,
  values: List[String],
  func: PushedFilterFun,
  expr: String)

class PrunePushedFilter extends Serializable {

  private val conditions = ListBuffer[PrunePushedFilterExpr]()

  def add(
    name: String,
    op: FilterOp,
    values: List[String],
    func: PushedFilterFun,
    expr: String
  ): PrunePushedFilter = {
    this.conditions += PrunePushedFilterExpr(name, op, values, func, expr)
    this
  }

  def filter(name: String): PrunePushedFilter = {
    val rs = this.conditions.filter(_.name == name)
    this.conditions.addAll(rs)
    this
  }

  def getNames: Set[String]                      = this.conditions.map(_.name).toSet
  def getConditions: List[PrunePushedFilterExpr] = this.conditions.toList

  override def toString: String = {
    this.conditions.map(_.expr).mkString("[", ",", "]")
  }
}
